### DISCUSSION TITLE
Strong types on resources' props

### DISCUSSION THREAD

**Discussion by leite08** [aws/aws-cdk#25688] [General]
Hi all! 👋 

I wish CDK types had stronger types on resources props - and searched on issues but couldn't find anything related.

Currently, one can define this and only learn its invalid at runtime, when stack is being deployed:

```ts
// non-fifo queue, as it doesn't specify `fifo: true`
new Queue(stack, name, {
  contentBasedDeduplication: aBooleanValue, // invalid, only applies to fifo queues
});
```

Typescript CDK could leverage the language's [discriminated unions](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#discriminated-unions) to avoid errors like the one above at compile time.

For example:

```ts
type AbstractQueueProps = {
  queueName?: string;
  retentionPeriod?: Duration;
  // ...other shared queue props
};

type StandardQueueProps = AbstractQueueProps & {
  // ...standard-specific props
  contentBasedDeduplication?: never;
  fifo?: never;
};

type FifoQueueProps = AbstractQueueProps & {
  fifo: true;
  contentBasedDeduplication?: boolean;
  // ...other fifo-specific props
};

type Queue = StandardQueueProps | FifoQueueProps;
```

I know the doc specifies `Only applies to FIFO queues`, but even better than doc is a compilation error, like so:

<img width="427" alt="image" src="https://github.com/aws/aws-cdk/assets/2132564/dc4f357a-6e14-4b72-9a73-c7ad3b178ebe">

<img width="963" alt="image" src="https://github.com/aws/aws-cdk/assets/2132564/cb3b3ea1-ffbe-4974-afd3-213812bb447c">

Does anyone know whether there's an issue to address this?

Is anyone interested in such improvement across the codebase?

**Edit**: fix the type structure introducing `AbstractQueueProps`

**Comment by peterwoodworth** [comment]
We can't use union types due to how jsii works to support other languages - however we do often validate props to ensure that the configurations will not conflict and cause failure at deployment. If there's a specific validation we can implement for Queue, we'd love to have an issue tracking that / a PR to fix that
